### DISCUSSION TITLE
Save FrameBuffer content to File?

### DISCUSSION THREAD

**Discussion by ingelumer** [micropython/micropython#10785] [General]
Hello,

I am running a SH1106 display on a Pico and would like to save the content to a file.

Does anybody know if it possible to save the FrameBuffer content to a file, eg as Bytearry or even as graphic (eg PNG/JPP)?

**Comment by mcauser** [comment]
When you initialise the FrameBuffer, you provide it a bytearray to store the contents.

Instead of passing the bytearray directly to the constructor:
```
fbuf = framebuf.FrameBuffer(bytearray(100 * 10 * 2), 100, 10, framebuf.RGB565)
```

Define one as a separate variable, then pass it in. This gives you a reference to the internal buffer and can easily write that to a file.
```
buf = bytearray(100 * 10 * 2)
fbuf = framebuf.FrameBuffer(buf, 100, 10, framebuf.RGB565)

with open('display.bin', 'wb') as f:
    f.write(buf)
```

FrameBuffer docs: https://docs.micropython.org/en/latest/library/framebuf.html

**Reply by ingelumer** [reply]
that worked great :-) thank you
But how can I convert the bin file to a readable image file?

**Reply by Josverl** [reply]
You'll need to clarify the question, readable by what or where?

**Reply by ingelumer** [reply]
readable in terms of creating the PBM file with the mentined code and opening the created PBM file in an ImageViewer and reading the new pbm through MicroPython code to a bytearray

**Comment by scruss** [comment]
PNG and JPEG might be a bit of a stretch, but if you're okay with something simpler, PBM (Portable Bitmap) is very easy to write, especially for mono OLEDs. The bytearray buffer just happens to match what PBM needs exactly. I use these routines, or variations on them:

``` python3
def readpbm(file):
    # returns framebuf read from file, width and height
    f = open(file, "rb")
    _ = f.readline()
    (w, h) = [int(x) for x in f.readline().decode().split()]
    b = bytearray(f.read())
    f.close()
    return (framebuf.FrameBuffer(b, w, h, framebuf.MONO_HLSB), w, h)


def writepbm(file, width, height, img_bytes):
    f = open(file, "wb")
    f.write("P4" + "\n")
    f.write(str(width) + " " + str(height) + "\n")
    f.write(img_bytes)
    f.close()
```

**Reply by mcauser** [reply]
A less than efficient way to get pngs: write a html page with some javascript to draw the mono byte data onto a html5 canvas, then right click on it and save as png 😄

**Reply by mcauser** [reply]
Or maybe just use the PIL (pillow) library. Might need a translation layer to massage the bytes into their desired format.

**Comment by ingelumer** [comment]
many thanks @scruss for your code and @mcauser for your suggestions.
Your code is creating a pbm file but the result is a scrambled picture in my case. 
Could this be caused by my SH1106 driver using VLSB instead of HLSB used in your example?

[disp2.zip](https://github.com/micropython/micropython/files/10861663/disp2.zip)

**Reply by scruss** [reply]
ah, yes - that'll be the cause. I didn't know that SH1106s used VLSB. Sorry about that

**Reply by ingelumer** [reply]
The conversion from VLSB to HLSB looks more difficult than expected.
I tried to convert the P4 PBM to P1 in order to have the binary values, transpose them from colums to rows and write the bits back to a new P1 file.
The result looks also pretty scrambled but far away from the content from my display.

Is my understanding correct that the order for VLSB is like below example?

01 05 09
02 06 10 
03 07 11
04 08 12

**Reply by scruss** [reply]
I think my read routine is still broken for VLSB, but here's a slow routine to write rather large Plain PBM files. I've tested it on HLSB and VLSB framebuffers.

``` python3
def writeplainpbm(file, width, height, framebuf):
    f = open(file, "wb")
    f.write("P1" + "\n")
    f.write(str(width) + " " + str(height) + "\n")
    for y in range(height):
        for x in range(width):
            f.write(str(framebuf.pixel(x, y)))
        f.write("\n")
    f.close()
```
By using [FrameBuffer.pixel()](https://docs.micropython.org/en/latest/library/framebuf.html?highlight=framebuf#framebuf.FrameBuffer.pixel) I've avoided messing with the framebuffer bytes at all.

**Comment by slabua** [comment]
What if some operations such as writing text or other primitives are performed without explicitly defining a framebuffer?

**Comment by slabua** [comment]
I think I got it. I modified my driver to return its bytearray, then I created a temporary framebuffer initialised with said array and, in my case, MONO_HMSB. Then used @scruss last function to write the pbm file and it seems to be working fine.
